pkg/operator/ceph/cluster/status/conditions.go

```diff
@@ -0,0 +1,300 @@
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionIgnored,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+func ConditionUpgrading(context *clusterd.Context, conditions *[]cephv1.RookConditions, reason, message string) {
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionNotReady,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionReady,
+		Status:  v1.ConditionFalse,
+		Reason:  "",
+		Message: "",
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionAvailable,
+		Status:  v1.ConditionFalse,
+		Reason:  "",
+		Message: "",
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionUpgrading,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+func ConditionFailure(context *clusterd.Context, conditions *[]cephv1.RookConditions, reason, message string) {
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionReady,
+		Status:  v1.ConditionFalse,
+		Reason:  "ClusterNotReady",
+		Message: "",
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionAvailable,
+		Status:  v1.ConditionFalse,
+		Reason:  "",
+		Message: "",
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionNotReady,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionFailure,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+func ConditionProgressing(context *clusterd.Context, conditions *[]cephv1.RookConditions, reason, message string) {
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionNotReady,
+		Status:  v1.ConditionTrue,
+		Reason:  "",
+		Message: message,
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionProgressing,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+func ConditionReady(context *clusterd.Context, conditions *[]cephv1.RookConditions, reason, message string) {
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionNotReady,
+		Status:  v1.ConditionFalse,
+		Reason:  "",
+		Message: message,
+	})
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionReady,
+		Status:  v1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+func ConditionUpdating(context *clusterd.Context, conditions *[]cephv1.RookConditions, reason, message string) {
+	setCondition(context, conditions, cephv1.RookConditions{
+		Type:    cephv1.ConditionReady,
+		Status:  v1.ConditionFalse,
```

Comment:
Again, even if we are updating the cluster, Ceph should still be ready. Will this be clear that it's just the operator that is not ready? I just want to reduce confusion about the difference between Ceph not being ready and Rook not being ready.

Reply:
Oh. Apart from the initialization of the cluster, can you point out an example of when will the cluster be Not Ready?

Reply:
I would expect that after the operator initially creates the cluster, the Ceph cluster is always in a Ready state. Otherwise, it means something in Ceph has failed and the Ceph status should show what the status is. What if we removed the conditions for ClusterReady and ClusterNotReady? They seem confusing with the ceph status.

Reply:
We can remove the ClusterNotReady condition. Cluster ready can be set to true once the cluster is created and then we will show the other conditions and we can set the `finalcondition` according to the error that happened on the cluster. Then I think we can just copy paste the `reason` and `message` field to all the other conditions so that the user will know the error has occured.

Reply:
Why would we copy them to all the other conditions? If there was a failure, wouldn't we just update that on the `ConditionFailure`?

Reply:
Yes. It would. Then I'll just remove the Not ready condition. I think keeping the Ready is not an issue. Anyway once the cluster is created it would be always there right. I'll just update other conditions accordingly